programs/clearing_house/src/math/fees.rs

```diff
@@ -103,19 +105,36 @@ pub fn calculate_fee_for_order_fulfill_against_amm(
 fn calculate_referrer_reward_and_referee_discount(
     fee: u128,
     fee_structure: &FeeStructure,
+    referrer_stats: &Option<&mut UserStats>,
 ) -> ClearingHouseResult<(u128, u128)> {
-    Ok((
-        get_proportion_u128(
-            fee,
-            fee_structure.referral_discount.referrer_reward_numerator,
-            fee_structure.referral_discount.referrer_reward_denominator,
-        )?,
-        get_proportion_u128(
-            fee,
-            fee_structure.referral_discount.referee_discount_numerator,
-            fee_structure.referral_discount.referee_discount_denominator,
-        )?,
-    ))
+    let max_referrer_reward_from_fee = get_proportion_u128(
+        fee,
+        fee_structure.referral_discount.referrer_reward_numerator,
+        fee_structure.referral_discount.referrer_reward_denominator,
+    )?;
+
+    let referrer_reward = match referrer_stats {
+        Some(referrer_stats) => {
+            let max_referrer_reward_in_epoch = cast_to_u128(
+                fee_structure
+                    .referral_discount
+                    .referrer_reward_epoch_upper_bound
+                    .checked_sub(referrer_stats.current_epoch_referrer_reward)
+                    .ok_or_else(math_error!())?
+                    .max(0),
```

Comment:
`referrer_reward_epoch_upper_bound` is u64 so this value can never be <0? Also might just do a `saturating_sub`